### PR TITLE
Alter Character Creation Process

### PR DESCRIPTION
**Pull request by michaelchessall**
Finalize your character through the creator, join game now only functions when you have a preexisting character.

Remove irrelevant menus and buttons from the character create screen

Force players to select a new character name before finalizing.

### CODE REVIEW

File: mods/persistence/modules/mob/new_player.dm [hunk 1]
```diff
@@ -155,3 +151,56 @@
...
+	if(!SSjobs.check_unsafe_spawn(src, spawn_turf))
+		return
+
+	SSjobs.assign_role(src, job.title, 1)
+
+	var/mob/living/character = create_character(spawn_turf)	//creates the human and transfers vars and mind
+	if(!character)
+		return 0
+	if(character.mind)
+		SSpersistence.AddToLimbo(character.mind, character.mind.unique_id, LIMBO_MIND, character.mind.key, TRUE)
```
Comment (PsyCommando) [inline mods/persistence/modules/mob/new_player.dm]: You don't need to copy paste and reimplement the whole proc just for this line. If you look at the spawnpoint code (/decl/spawnpoint/chargen/after_join()), you can have the character save after there instead. It'll make a lot less problem in the future for us.

File: mods/persistence/modules/client/preferences.dm [hunk 2]
```diff
@@ -0,0 +1,505 @@
+/* PLACEHOLDER VERB UNTIL SAVE INIT (or whatever the issue is) IS FIXED */
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preferences.dm]: You shouldn't copy the whole thing over if you're just overriding one or two procs. Its going to be a pain in the ass to maintain and fix bugs, and its just not good practice.

File: mods/persistence/modules/client/preference_setup/preference_setup.dm [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+/datum/category_group/player_setup_category/aspect_preferences
+	name = "Aspects"
+	sort_order = 3
+	category_item_type = /datum/category_item/player_setup_item/aspects
+
+/datum/category_group/player_setup_category/background_preferences/content(var/mob/user)
+	. = ""
+	for(var/datum/category_item/player_setup_item/PI in items)
+		. += "[PI.content(user)]<br>"
+/**
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preference_setup/preference_setup.dm]: First, again, you don't want to copy the whole thing just for changing one or two things.

And then, you can't do that. Its going to break preferences access down the line, I know I tried that before. If you want to remove elements from the menu, its going to need more than just commenting them like that.

File: nebula.dme [hunk 4]
```diff
@@ -1548,25 +1549,21 @@
 #include "code\modules\client\client_helpers.dm"
 #include "code\modules\client\client_procs.dm"
 #include "code\modules\client\movement.dm"
-#include "code\modules\client\preferences.dm"
```
Comment (PsyCommando) [inline nebula.dme]: Ideally we don't want to remove includes from nebula.dme. And you don't really need to with your current changes either?

File: nebula.dme [hunk 5]
```diff
@@ -3460,5 +3457,9 @@
 #include "maps\~mapsystem\maps_unit_testing.dm"
 #include "maps\~unit_tests\unit_testing.dm"
 #include "mods\_modpack.dm"
+#include "mods\persistence\modules\client\preferences.dm"
```
Comment (PsyCommando) [inline nebula.dme]: Don't include modpack stuff in here. Things in the modpack have to go under mods/presistence/persistence.dme otherwise things are gonna get really wonky with nebula.dme being loaded before persistence.dme.

**Review by PsyCommando — CHANGES_REQUESTED** [review]
Its hard to tell what you've changed since you're copy pasting a lot of code. If you can just copy-paste the procs you need to override it would make this a lot easier to review, and also its just straight up better practice not to copy paste entire files just to change a proc or two.

Also, the way you're just commenting out entire sections of the prefs without modifying the places where the game accesses the prefs you've removed is going to cause issues. The way prefs work is very hardcoded. You need all the prefs categories to load, or stuff will runtime. There's probably a way to override what's actually displayed in char setup and what's not in some of the procs though. 
Its probably easier to just rewrite the whole thing, and refer to the bits we need from the base ui.

File: mods/persistence/modules/client/preference_setup/preference_setup.dm [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+	for(var/category_type in typesof(category_group_type))
+		if(category_type in preferences.skipped_menus)
+			continue
+		var/datum/category_group/category = category_type
+		if(initial(category.name))
+			category = new category(src)
+			categories += category
+			categories_by_name[category.name] = category
+	categories = sortTim(categories, /proc/cmp_category_groups)
+	selected_category = categories[1]
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preference_setup/preference_setup.dm]: You should override the header() proc instead. Because, if you prevent the categories from being added to the the category list then we have the same problem we had before, because all categories need to init their preference field.
If you just skip drawing the category button in the header like I'm suggesting, nobody can select the categories and they still do their thing in the background.

```suggestion
/datum/category_collection/player_setup_collection/proc/header()
	var/dat = ""
	for(var/datum/category_group/player_setup_category/PS in categories)
		if(is_type_in_list(PS, hidden_categories))
			continue //Skip categories we don't wanna display
		if(PS == selected_category)
			dat += "[PS.name] "	// TODO: Check how to properly mark a href/button selected in a classic browser window
		else
			dat += "<a href='?src=\ref[src];category=\ref[PS]'>[PS.name]</a> "
	return dat
```

The hidden_category list is just paths to the category class we don't want to display.

File: mods/persistence/_persistence.dme [hunk 7]
```diff
@@ -1,19 +1,19 @@
 #ifndef MODPACK_PERSISTENCE
```
Comment (PsyCommando) [inline mods/persistence/_persistence.dme:1]: I suggest you revert your persistence.dme to before you restored it, because the diff is acting funky with all the lines being moved around, I had to run winmerge on it to make sure what was going on. 
Also, seems like all the files you added under "modules/client/" aren't included in the persistence.dme. They won't get compiled if you don't include them.

**Review by PsyCommando — CHANGES_REQUESTED** [review]

File: mods/persistence/_persistence.dme (hunk 7 above)
Comment (michaelchessall) [reply mods/persistence/_persistence.dme:1]: frick

File: mods/persistence/modules/client/preference_setup/preference_setup.dm [hunk 8]
```diff
@@ -0,0 +1,10 @@
+/datum/category_collection/player_setup_collection/proc/header()
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preference_setup/preference_setup.dm]: You're just missing this: 
```suggestion
/datum/category_collection/player_setup_collection
	var/static/list/hidden_categories = list(\
		/datum/category_group/player_setup_category/occupation_preferences,\
		/datum/category_group/player_setup_category/record_preferences, \
		/datum/category_group/player_setup_category/law_pref,\
		/datum/category_group/player_setup_category/controls,\
		/datum/category_group/player_setup_category/global_preferences,\
	)

/datum/category_collection/player_setup_collection/proc/header()
```

Comment (PsyCommando) [reply mods/persistence/modules/client/preference_setup/preference_setup.dm]: Add or remove the fields you want or not.

**Review by PsyCommando — CHANGES_REQUESTED** [review]

File: mods/persistence/modules/client/preference_setup/preference_setup.dm [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/datum/category_collection/player_setup_collection
+	var/static/list/hidden_categories = list(\
+		/datum/category_group/player_setup_category/occupation_preferences,\
+		/datum/category_group/player_setup_category/record_preferences, \
+		/datum/category_group/player_setup_category/law_pref,\
+		/datum/category_group/player_setup_category/controls,\
+		/datum/category_group/player_setup_category/global_preferences,\
+	)
+
+/datum/category_collection/player_setup_collection/proc/header()
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preference_setup/preference_setup.dm]: woops
```suggestion
/datum/category_collection/player_setup_collection/header()
```

File: mods/persistence/modules/client/preferences.dm [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+			to_chat(usr, "<span class='danger'>The must set a unique character name to continue.</span>")
+			return
+		save_preferences()
+		save_character()
+		switch(alert("Are you sure you want to finalize your character and join the game with the character you've created?", "Character Confirmation", "Yes", "No"))
+			if("No")
+				return
+		if(isnewplayer(client.mob))
+			var/mob/new_player/M = client.mob
+			M.AttemptLateSpawnOutreach(SSjobs.get_by_path(global.using_map.default_job_type))
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preferences.dm]:
```suggestion
			AttemptLateSpawn(SSjobs.get_by_path(global.using_map.default_job_type))
```

File: mods/persistence/modules/client/preferences.dm [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+			to_chat(usr, "<span class='danger'>The must set a unique character name to continue.</span>")
+			return
+		save_preferences()
+		save_character()
+		switch(alert("Are you sure you want to finalize your character and join the game with the character you've created?", "Character Confirmation", "Yes", "No"))
+			if("No")
+				return
+		if(isnewplayer(client.mob))
+			var/mob/new_player/M = client.mob
+			AttemptLateSpawn(SSjobs.get_by_path(global.using_map.default_job_type))
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preferences.dm]:
```suggestion
			M.AttemptLateSpawn(SSjobs.get_by_path(global.using_map.default_job_type))
```

**Comment by PsyCommando** [comment]
I think I fixed all the little hiccups, but you should test it now. If you confirm it all works, I'll merge it then.

**Comment by michaelchessall** [comment]
alright its working good now, everything should be okay and lets never speak of all the trouble ive had with this again 💯

File: mods/persistence/modules/client/preferences.dm [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+	"}
+	charpanel.set_content(content)
+	charpanel.open() // Skip registring onclose on the browser panel
+
+/datum/preferences/Topic(href, list/href_list)
+	if(..())
+		return TRUE
+	if(href_list["finish"])
+		if(!real_name)
+			to_chat(usr, "<span class='danger'>The must set a unique character name to continue.</span>")
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preferences.dm:62]: The string has some typos here.

File: mods/persistence/modules/client/preference_setup/preference_setup.dm [hunk 13]
```diff
@@ -0,0 +1,20 @@
+/datum/category_collection/player_setup_collection
+	var/static/list/hidden_categories = list(\
+		/datum/category_group/player_setup_category/appearance_preferences,\
```
Comment (PsyCommando) [inline mods/persistence/modules/client/preference_setup/preference_setup.dm:3]: Just wondering but why did you add appearence_preferences there?

**Review by PsyCommando — APPROVED** [review]
There's like one typo and some things I'm not too sure about, but worst case we can patch that. I'll approve it.
